File: ft_yacc/template/template.c
/* CODE_BEFORE */
#include <stdio.h>
#include <stdlib.h>

/* REMOVE */
#include "../calc.tab.h"
/* REMOVE */

int yychar = -2;

/* DEFINES */

#define STACK_CAP 100
#define yyerrok (yy_recovering = 0)
#define yyclearin (yychar = YYEMPTY)

#define YYACCEPT goto yyacceptlabel
#define YYABORT goto yyabortlabel
#define YYERROR goto yyerrorlabel

#define YYRECOVERING() (!!yy_recovering)

enum yysymbol_kind_t {
	YYSYMBOL_YYEMPTY = -2,
	/* REMOVE */
	YYSYMBOL_END = 0,
	YYSYMBOL_ERROR = 3,
	/* REMOVE */
	/* TOKENS */
};
typedef enum yysymbol_kind_t yysymbol_kind_t;

/* TABLES */

/* REMOVE */
static const ssize_t yy_gotos[0][0] = {};
static const ssize_t yy_indexes[0] = {};
static const size_t yy_rlens[0] = {};
static const size_t yy_products[0] = {};
static const ssize_t yy_actions[0] = {};
static const ssize_t yy_default_act[0] = {};
static const ssize_t yy_default_reduce_act[0] = {};

/* REMOVE */

/* DEBUGGING_TABLES */
/* REMOVE */
const static int yy_lines[] = {};
const static int yy_terminal[] = {};
const static char *const yy_names[] = {};

/* REMOVE */

typedef struct {
	size_t state;
	YYSTYPE value;
#if YYDEBUG
	ssize_t id;
#endif
} yyparse_elt;

YYSTYPE yylval;
static yyparse_elt *yyparse_stack = NULL;
static yyparse_elt *yyparse_stack_top = NULL;
static size_t yyparse_stack_cap = STACK_CAP;
static YYSTYPE yychar_value;
static ssize_t yychar_id;
static int yyact;
static size_t yylen;
static YYSTYPE yyval;
static int yy_recovering = 0;
static int yytoken_since_error = 0;

#ifndef yydebug
int yydebug;
#endif

#ifndef yyerror
void yyerror(const char *);
#endif

int yyparse(void);

#ifndef yylex
int yylex(void);
#endif

// ----- error --------------------

void yyexiterror(const char *msg) {
	fprintf(stderr, "%s\n", msg);
	free(yyparse_stack);
	yyparse_stack = NULL;
	exit(1);
}

// ----- debug --------------------

#if YYDEBUG
void yylog(const char *fmt, ...) {
	if (!yydebug) {
		return;
	}
	va_list ap;
	va_start(ap, fmt);
	vfprintf(stderr, fmt, ap);
	va_end(ap);
}

const char *token_class(size_t token_id) {
	return yy_terminal[token_id] ? "token" : "nterm";
}

void print_stack() {
	if (!yydebug) {
		return;
	}
	fprintf(stderr, "Stack now ");
	yyparse_elt *it = yyparse_stack;
	while (it < yyparse_stack_top) {
		fprintf(stderr, "%zu", it->state);
		it++;
		if (it < yyparse_stack_top + 1)
			fprintf(stderr, ", ");
	}
	fprintf(stderr, "\n");
}
#endif

// ----- parse stack management --------------------

size_t yyparse_stack_len() {
	return yyparse_stack_top - yyparse_stack + 1;
}

void yyparse_stack_drain(size_t len) {
	if (yyparse_stack_len() < len) {
		yyexiterror("parse stack underflow");
	}
	yyparse_stack_top -= len;
}

void yyparse_stack_pop() {
	yyparse_stack_drain(1);
}

void yyparse_stack_push_eps(int state, ssize_t token_id) {
	if (yyparse_stack_len() >= yyparse_stack_cap) {
		yyparse_stack_cap += STACK_CAP;
		size_t len = yyparse_stack_len() - 1;
		yyparse_stack = realloc(yyparse_stack, yyparse_stack_cap * sizeof(yyparse_elt));
		yyparse_stack_top = yyparse_stack + len;
		if (!yyparse_stack) {
			yyexiterror("alloc failed");
		}
	}
	yyparse_stack_top++;
	yyparse_stack_top->state = state;
#if YYDEBUG
	yyparse_stack_top->id = token_id;
	yylog("Entering state %zu\n", state);
	print_stack();
#endif
}

void yyparse_stack_push(int state, YYSTYPE value, ssize_t token_id) {
	yyparse_stack_push_eps(state, token_id);
	yyparse_stack_top->value = value;
}

// ----- parser actions --------------------

void yyread_token() {
	yychar = yylex();
	if (yychar <= 0) {
		yychar = YYEOF;
	}
	yychar_id = yy_indexes[yychar];
	yychar_value = yylval;

#if YYDEBUG
	yylog("Reading a token\n");
	if (yychar == 0) {
		yylog("Now at end of input.\n");
	} else {
		yylog("Next token is token %s\n", yy_names[yychar_id]);
	}
#endif
}

void yyshift() {
	yyact += 1;

	if (yy_recovering && yytoken_since_error >= 2) {
		yy_recovering = 0;
#if YYDEBUG
		yylog("Error recovered\n");
#endif
	} else if (yy_recovering) {
		yytoken_since_error += 1;
	}

	if (yychar == YYEMPTY)
		yyread_token();
#if YYDEBUG
	yylog("Shifting token %s\n", yy_names[yychar_id]);
#endif
	yyparse_stack_push(-yyact, yychar_value, yychar_id);
	yychar = YYEMPTY;
	yychar_id = YYSYMBOL_YYEMPTY;
}

int yynext_act() {
	size_t state_id = yyparse_stack_top->state;
	ssize_t dflt = yy_default_act[state_id];
	if (dflt)
		return dflt;
	if (yychar == YYEMPTY)
		yyread_token();
	dflt = yy_gotos[state_id][yychar_id];
	if (dflt)
		return dflt;
	return yy_default_reduce_act[state_id];
}

// ----- error --------------------

int yyunwind() {
	while (yyparse_stack_len() > 0 && yy_gotos[yyparse_stack_top->state][YYSYMBOL_ERROR] >= 0) {
#if YYDEBUG
		yylog(
			"Error: popping %s %s\n",
			token_class(yyparse_stack_top->id),
			yy_names[yyparse_stack_top->id]);
		print_stack();
#endif
		yyparse_stack_pop();
	}
	if (yyparse_stack_len() == 0) {
		return 1;
	}
	yy_recovering = 1;
	yyact = yy_gotos[yyparse_stack_top->state][YYSYMBOL_ERROR] + 1;
#if YYDEBUG
	yylog("Shifting token error\n");
#endif
	yyparse_stack_push_eps(-yyact, YYSYMBOL_ERROR);
	yytoken_since_error = 0;
	yyact = yy_gotos[yyparse_stack_top->state][yychar_id];
	while (yyact == 0 && yychar != YYEOF) {
		yyclearin;
		yyread_token();
		yyact = yy_gotos[yyparse_stack_top->state][yychar_id];
	}
	return 0;
}

// ----- yyparse --------------------

int yyparse() {
#if YYDEBUG
	yylog("Start parsing\n");
#endif

	yyparse_stack = malloc(yyparse_stack_cap * sizeof(yyparse_elt));
	yyparse_stack_top = yyparse_stack;
	if (!yyparse_stack) {
		yyexiterror("alloc failed");
	}

	yyparse_stack_push_eps(0, 0);

	yyread_token();

	while (1) {
		yyact = yynext_act();
		if (yyact == 1) {
			break;
		}
		if (yyact > 0) {

			yyact -= 1;
#if YYDEBUG
			yylog("Reducing stack by rule %d (line %d)\n", yyact, yy_lines[yyact]);
#endif
			size_t len = yy_rlens[yyact];
			size_t product_id = yy_products[yyact];
			yylen = yy_rlens[yyact];
			yyval = (yyparse_stack_top)[1 - yylen].value;
#if YYDEBUG
			int i = 0;
			while (i < len) {
				yyparse_elt *token = yyparse_stack_top - i;
				yylog("   $%d = %s %s\n", i + 1, token_class(token->id), yy_names[token->id]);
				i++;
			}
			yylog("-> $$ = %s %s\n", token_class(product_id), yy_names[product_id]);
#endif

			switch (yy_actions[yyact]) {
				/* ACTIONS */

			default:
				break;
			}

			yyparse_stack_drain(yylen);
			yyparse_stack_push(
				-(yy_gotos[yyparse_stack_top->state][yy_products[yyact]] + 1), yyval, product_id);
		} else if (yyact < 0) {
			yyshift();
		} else {
			if (yy_recovering == 0)
				yyerror("syntax error");
		yyerrorlabel:
			if (yy_recovering == 0) {
				if (yyunwind())
					YYABORT;
			} else if (yychar == YYEOF)
				YYABORT;
			else
				yyclearin;
		}
	}
#if YYDEBUG
	while (yyparse_stack != yyparse_stack_top) {
		yyparse_elt *token = yyparse_stack_top;
		yylog("Cleanup: popping %s %s\n", token_class(token->id), yy_names[token->id]);
		yyparse_stack_pop();
	}
#endif

yyacceptlabel:
	return 0;
yyabortlabel:
	return 3;
}

File: ft_yacc/template/template.rs
#![allow(unused_braces, mixed_script_confusables, unused)]
use std::fmt;
use std::io::Read;

/* CODE_BEFORE */

/* TOKENS */

/* FEEDBACK */
pub trait YYFeedback {
    fn yy_feedback(&mut self, accepts: &dyn Fn(usize) -> bool);
}
/* FEEDBACK */

pub struct Yacc<R: Read> {
    state_stack: Vec<usize>,
    value_stack: Vec<YYToken>,
    span_stack: Vec<Span>,
    lookahead: Option<YYToken>,
    lookahead_id: usize,
    lookahead_span: Span,
    act: isize,
    continue_parse: bool,
    is_recovering: bool,
    token_since_error: usize,
    yy_status: i32,
    pub span: Span,
    pub yydebug: bool,
    pub lexer: YYLex<R>,
}

/* DEBUGGING */
macro_rules! yylog {
    ($self:expr, $($arg:tt)*) => {
        if $self.yydebug {
            eprintln!($($arg)*);
        }
    };
}
/* DEBUGGING */

impl<R: Read> Yacc<R> {
    /* REMOVE */
    const YY_EOF_TOKEN_ID: usize = 2;
    const YY_ERROR_TOKEN_ID: usize = 3;
    const YY_GOTO_TABLE: [[isize; 0]; 0] = [];
    const YY_RLEN_TABLE: [usize; 0] = [];
    const YY_DEFAULT_ACT: [usize; 0] = [];
    /* REMOVE */

    /* TABLES */
    /* DEBUGGING_TABLES */

    pub fn new(lexer: YYLex<R>) -> Self {
        Self {
            state_stack: Vec::new(),
            value_stack: Vec::new(),
            span_stack: Vec::new(),
            span: Span::default(),
            lookahead: None,
            lookahead_id: 0,
            lookahead_span: Span::default(),
            yydebug: false,
            act: 0,
            continue_parse: true,
            is_recovering: false,
            lexer,
            token_since_error: 0,
            yy_status: 0,
            /* DEFINES */
        }
    }

    /* FEEDBACK */
    fn yy_accepts_at(state_stack: &[usize], token_id: usize) -> bool {
        let mut stack = state_stack.to_vec();
        for _ in 0..Self::YY_PROBE_LIMIT {
            let &state = stack.last().unwrap();
            let mut act = Self::YY_DEFAULT_ACT[state];
            if act == 0 {
                act = Self::YY_GOTO_TABLE[state][token_id];
            }
            if act == 0 {
                act = Self::YY_DEFAULT_REDUCE_ACT[state];
            }
            match act {
                1 => return true,
                a if a < 0 => return true,
                0 => return false,
                _ => (),
            }
            let rule = (act - 1) as usize;
            let len = Self::YY_RLEN_TABLE[rule];
            if len >= stack.len() {
                return false;
            }
            stack.truncate(stack.len() - len);
            let &top = stack.last().unwrap();
            let goto = Self::YY_GOTO_TABLE[top][Self::YY_PRODUCT_TABLE[rule]];
            if goto >= 0 {
                return false;
            }
            stack.push(-(goto + 1) as usize);
        }
        false
    }

    const YY_PROBE_LIMIT: usize = 1024;
    /* FEEDBACK */

    fn next_act(&mut self) -> isize {
        let &state_id = self.state_stack.last().unwrap();
        let mut dflt = Self::YY_DEFAULT_ACT[state_id];
        if dflt != 0 {
            return dflt;
        }
        if self.lookahead.is_none() {
            self.read_token();
        }
        dflt = Self::YY_GOTO_TABLE[state_id][self.lookahead_id];
        if dflt != 0 {
            return dflt;
        }
        Self::YY_DEFAULT_REDUCE_ACT[state_id]
    }

    pub fn yyparse(&mut self) -> i32 {
        /* DEBUGGING */
        yylog!(self, "Starting parse");
        /* DEBUGGING */

        self.push_statcks(0, YYToken::yyeof, Span::default());
        self.read_token();

        while self.continue_parse {
            let state_id = self.state_stack.last().unwrap();
            self.act = self.next_act();
            match self.act {
                1 => break,
                a if a > 0 => self.reduce(),
                a if a < 0 => self.shift(),
                _ => self.error(),
            }
        }
        /* DEBUGGING */
        for token in self.value_stack[1..].iter().rev() {
            yylog!(self, "Cleanup: popping {:?}", token);
        }
        /* DEBUGGING */
        self.yy_status
    }

    fn error(&mut self) {
        if !self.is_recovering {
            self.unwind()
        } else if matches!(self.lookahead, Some(YYToken::yyeof)) {
            self.yyabort();
        } else {
            self.yyclearin();
        }
    }

    fn unwind(&mut self) {
        yyerror(self.error_message(), self);
        let start = self.span_stack.last().cloned().unwrap_or_default().start;
        while let Some(i) = self.state_stack.last()
            && Self::YY_GOTO_TABLE[*i][Self::YY_ERROR_TOKEN_ID] == 0
        {
            self.state_stack.pop();
            self.span_stack.pop();
            let val = self.value_stack.pop().unwrap();
            /* DEBUGGING */
            yylog!(self, "Error: popping {:?}", val);
            yylog!(self, "Stack now {:?}", self.state_stack);
            /* DEBUGGING */
        }
        let end = self.span_stack.last().cloned().unwrap_or_default().end;

        let Some(i) = self.state_stack.last() else {
            self.yyabort();
            return;
        };
        self.is_recovering = true;
        let act = Self::YY_GOTO_TABLE[*i][Self::YY_ERROR_TOKEN_ID] + 1;

        /* DEBUGGING */
        yylog!(self, "Shifting token {:?}", YYToken::error);
        /* DEBUGGING */

        self.push_statcks((-act) as usize, YYToken::error, Span::new(start, end));
        self.token_since_error = 0;

        let top = (-act) as usize;
        self.act = Self::YY_GOTO_TABLE[top][self.lookahead_id];
        while self.act == 0 {
            self.yyclearin();
            self.read_token();
            if matches!(self.lookahead, Some(YYToken::yyeof)) {
                return;
            }
            self.act = Self::YY_GOTO_TABLE[top][self.lookahead_id];
        }
    }

    fn error_message(&self) -> String {
        "syntax error".to_string()
    }

    pub fn yyerrok(&mut self) {
        self.is_recovering = false;
    }

    pub fn yyaccept(&mut self) {
        self.continue_parse = false;
    }

    pub fn yyabort(&mut self) {
        self.yy_status = 3;
        self.continue_parse = false;
    }

    pub fn yyrecovering(&mut self) -> bool {
        self.is_recovering
    }

    pub fn yyclearin(&mut self) {
        self.lookahead = None;
    }

    fn push_statcks(&mut self, state: usize, value: YYToken, span: Span) {
        self.state_stack.push(state);
        self.value_stack.push(value);
        self.span_stack.push(span);
        /* DEBUGGING */
        yylog!(self, "Entering state {}", state);
        yylog!(self, "Stack now {:?}", self.state_stack)
        /* DEBUGGING */
    }

    fn read_token(&mut self) {
        /* FEEDBACK */
        let state_stack = &self.state_stack;
        self.lexer
            .yy_feedback(&|token_id| Self::yy_accepts_at(state_stack, token_id));
        /* FEEDBACK */
        let c = self.lexer.yylex();
        self.lookahead_id = c.index();
        self.lookahead_span = self.lexer.span;
        self.lookahead = Some(c);
        /* DEBUGGING */
        yylog!(self, "Reading a token");
        match &self.lookahead {
            None => yylog!(self, "Now at end of input."),
            Some(l) => yylog!(self, "Next token is token {:?}", l),
        }
        /* DEBUGGING */
    }

    fn shift(&mut self) {
        self.act += 1;

        if self.is_recovering && self.token_since_error >= 2 {
            self.is_recovering = false;
        } else if self.is_recovering {
            self.token_since_error += 1;
        }

        if self.lookahead.is_none() {
            self.read_token();
        }

        /* DEBUGGING */
        yylog!(
            self,
            "Shifting token {:?}",
            self.lookahead.as_ref().unwrap_or(&YYToken::yyeof)
        );
        /* DEBUGGING */
        let token = self.lookahead.take().unwrap_or(YYToken::yyeof);
        self.push_statcks(-self.act as usize, token, self.lexer.span);
    }

    fn reduce(&mut self) {
        self.act -= 1;

        /* DEBUGGING */
        yylog!(
            self,
            "Reducing stack by rule {} (line {}):",
            self.act,
            Self::YY_PRODUCTION_LINE[self.act as usize]
        );
        /* DEBUGGING */
        let len = Self::YY_RLEN_TABLE[self.act as usize];
        let product = Self::YY_PRODUCT_TABLE[self.act as usize];

        let yyval = self.action();
        if yyval == YYToken::error && Self::YY_ACTION_TABLE[self.act as usize] != -1 {
            self.error();
            return;
        }

        let &top = self.state_stack.last().unwrap();
        self.push_statcks(-(Self::YY_GOTO_TABLE[top][product] + 1) as usize, yyval, self.span);
    }

    #[allow(unused_braces, clippy::let_and_return)]
    fn action(&mut self) -> YYToken {
        let len = self.value_stack.len();
        let idx = len - Self::YY_RLEN_TABLE[self.act as usize];

        if idx == len {
            let pos = self.span_stack[len - 1].end;
            self.span.start = pos;
            self.span.end = pos;
        } else {
            self.span.start = self.span_stack[idx].start;
            self.span.end = self.span_stack[len - 1].end;
        }

        /* DEBUGGING */
        for (i, tok) in self.value_stack[idx..].iter().enumerate().rev() {
            yylog!(self, "   ${} = {:?}", i + 1, tok);
        }
        /* DEBUGGING */

        let ret = self.do_action(idx);
        self.value_stack.truncate(idx);
        self.state_stack.truncate(idx);
        self.span_stack.truncate(idx);

        /* DEBUGGING */
        yylog!(self, "-> $$ = {:?}", ret);
        /* DEBUGGING */
        ret
    }

    fn do_action(&mut self, idx: usize) -> YYToken {
        let rlen = Self::YY_RLEN_TABLE[self.act as usize];
        match Self::YY_ACTION_TABLE[self.act as usize] {
            /* ACTIONS */
            -1 => {
                if rlen == 0 {
                    YYToken::Empty
                } else {
                    std::mem::replace(&mut self.value_stack[idx], YYToken::Empty)
                }
            }
            _ => unreachable!(),
        }
    }
}

/* MAIN */
pub fn yyerror<D: fmt::Display, R: Read>(msg: D, yacc: &Yacc<R>) {
    eprintln!("{}", msg);
}

fn main() {
    use std::process::exit;

    let mut lexer = YYLex::default();
    let mut yacc = Yacc::new();
    exit(yacc.yyparse());
}
